lib/sales_engine.rb

```diff
@@ -153,56 +145,68 @@ def find_invoice_from_transaction(id)
     invoice_repository.find_by_id(id)
   end
 
-
-  def find_successful_invoices_from_merchant(id, date)
+  def find_invoice_ids_by_merchant_and_date(merchant_id, date)
     all_invoices = merchant_repository.find_invoices_by_merchant(id)
     if date == "all"
       invoice_ids = all_invoices.map(&:id)
     else
       invoice_ids = all_invoices.map do |invoice|
-        invoice.id unless invoice.created_at != date
+        invoice.id if invoice.created_at == date
       end
     end
+  end
 
-    all_transactions = invoice_ids.map do |invoice_id|
+  def find_transactions_from_invoice_ids(invoice_ids)
+    invoice_ids.map do |invoice_id|
       transaction_repository.find_all_by_invoice_id(invoice_id)
     end
+  end
 
-    successful_ids = all_transactions.flatten.map do |transaction|
-      transaction.invoice_id unless transaction.result != ("success")
+  def find_successful_invoice_ids_from_transactions(transactions)
+    transactions.flatten.map do |transaction|
+      transaction.invoice_id if transaction.result == ("success")
     end
+  end
 
-    invoice_items = successful_ids.map do |invoice_id|
+  def invoice_items_from_invoice_ids(invoice_ids)
+    invoice_ids.map do |invoice_id|
       invoice_item_repository.find_all_by_invoice_id(invoice_id)
     end
-    return (invoice_items)
+  end
+
+  def find_successful_invoice_items_from_merchant(id, date)
+    invoice_ids = find_invoice_ids_by_merchant_and_date(id, date)
+    all_transactions = find_transactions_from_invoice_ids(invoice_ids)
+    invoices = find_successful_invoice_ids_from_transactions(all_transactions)
+    invoice_items_from_invoice_ids(invoices)
   end
 
   def find_revenue_from_merchant(id, date)
     revenue = 0
-    find_successful_invoices_from_merchant(id, date).flatten.each do |invoice_item|
+    find_successful_invoice_items_from_merchant(id, date).flatten.each do |invoice_item|
         revenue = revenue + invoice_item.unit_price * invoice_item.quantity
     end
     revenue
   end
 
   def find_count_items_from_merchant(id, date="all")
-    count = find_successful_invoices_from_merchant(id, date).flatten.reduce(0) do |count, invoice_item|
+    count = find_successful_invoice_items_from_merchant(id, date).flatten.reduce(0) do |count, invoice_item|
```

Comment:
Shadowing outer local variable - `count`.<br>Line is too long. [108/80]<br>Prefer `inject` over `reduce`.